File: javascript/packages/linter/src/rules/erb-no-unused-local-variable.ts
import { ParserRule } from "../types.js"
import { StateScopeMap } from "../utils/state-directives-utils.js"
import { PrismVisitor, isPrismNodeType, locationFromByteOffset } from "@herb-tools/core"

import type { ParseResult, ParserOptions, PrismNodes } from "@herb-tools/core"
import type { FullRuleConfig, LintContext, UnboundLintOffense } from "../types.js"

const IGNORED_PREFIX = "_"

interface Scope {
  readonly writes: PrismNodes.LocalVariableWriteNode[]
  readonly referenced: Set<string>
}

class LocalVariableCollector extends PrismVisitor {
  readonly argumentWrites = new Set<PrismNodes.LocalVariableWriteNode>()

  private readonly scopes: Scope[] = []
  private readonly stack: Scope[] = []

  get unusedWrites(): PrismNodes.LocalVariableWriteNode[] {
    const unused = this.scopes.flatMap(scope => scope.writes.filter(write => !scope.referenced.has(write.name)))

    return unused.sort((first, second) => first.nameLoc.startOffset - second.nameLoc.startOffset)
  }

  override visitProgramNode(node: PrismNodes.ProgramNode): void {
    this.withScope(() => this.visitChildNodes(node))
  }

  override visitBlockNode(node: PrismNodes.BlockNode): void {
    this.withScope(() => this.visitChildNodes(node))
  }

  override visitLambdaNode(node: PrismNodes.LambdaNode): void {
    this.withScope(() => this.visitChildNodes(node))
  }

  override visitDefNode(node: PrismNodes.DefNode): void {
    this.withScope(() => this.visitChildNodes(node))
  }

  override visitClassNode(node: PrismNodes.ClassNode): void {
    this.withScope(() => this.visitChildNodes(node))
  }

  override visitModuleNode(node: PrismNodes.ModuleNode): void {
    this.withScope(() => this.visitChildNodes(node))
  }

  override visitSingletonClassNode(node: PrismNodes.SingletonClassNode): void {
    this.withScope(() => this.visitChildNodes(node))
  }

  override visitCallNode(node: PrismNodes.CallNode): void {
    for (const argument of node.arguments_?.arguments_ ?? []) {
      if (isPrismNodeType(argument, "LocalVariableWriteNode")) this.argumentWrites.add(argument)
    }

    this.visitChildNodes(node)
  }

  override visitLocalVariableWriteNode(node: PrismNodes.LocalVariableWriteNode): void {
    if (!node.name.startsWith(IGNORED_PREFIX)) this.scopeAt(node.depth)?.writes.push(node)

    this.visitChildNodes(node)
  }

  override visitLocalVariableReadNode(node: PrismNodes.LocalVariableReadNode): void {
    this.reference(node.name, node.depth)

    this.visitChildNodes(node)
  }

  override visitLocalVariableOperatorWriteNode(node: PrismNodes.LocalVariableOperatorWriteNode): void {
    this.reference(node.name, node.depth)

    this.visitChildNodes(node)
  }

  override visitLocalVariableAndWriteNode(node: PrismNodes.LocalVariableAndWriteNode): void {
    this.reference(node.name, node.depth)

    this.visitChildNodes(node)
  }

  override visitLocalVariableOrWriteNode(node: PrismNodes.LocalVariableOrWriteNode): void {
    this.reference(node.name, node.depth)

    this.visitChildNodes(node)
  }

  private withScope(visitChildren: () => void): void {
    const scope: Scope = { writes: [], referenced: new Set<string>() }

    this.scopes.push(scope)
    this.stack.push(scope)

    visitChildren()

    this.stack.pop()
  }

  private reference(name: string, depth: number): void {
    this.scopeAt(depth)?.referenced.add(name)
  }

  private scopeAt(depth: number): Scope | null {
    return this.stack[this.stack.length - 1 - depth] ?? null
  }
}

export class ERBNoUnusedLocalVariableRule extends ParserRule {
  static ruleName = "erb-no-unused-local-variable"
  static introducedIn = this.version("unreleased")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: {
        cli: "error",
        editor: "info",
      },
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return {
      prism_program: true,
    }
  }

  check(result: ParseResult, _context?: Partial<LintContext>): UnboundLintOffense[] {
    const program = result.value.prismNode
    const source = result.value.source

    if (!program) return []
    if (!source) return []

    const collector = new LocalVariableCollector()

    collector.visit(program)

    const stateNames = new Set(StateScopeMap.collect(result.value).allNames())

    return collector.unusedWrites.filter(write => !stateNames.has(String(write.name))).map(write => {
      const { startOffset, length } = write.nameLoc

      return this.createOffense(
        `Local variable \`${write.name}\` is assigned but never used. ${this.adviceFor(collector, write, write.name)}`,
        locationFromByteOffset(source, startOffset, length),
        undefined,
        undefined,
        ["unnecessary"],
      )
    })
  }

  private adviceFor(collector: LocalVariableCollector, write: PrismNodes.LocalVariableWriteNode, name: string): string {
    if (collector.argumentWrites.has(write)) {
      return `This assignment sits in an argument list, where it still passes the value positionally. Write \`${name}:\` if you meant a keyword argument, or drop the \`${name} =\`.`
    }

    return `Remove the assignment, or prefix it with an underscore as \`_${name}\` to show it is intentionally unused.`
  }
}

File: rust/herb-linter/src/rules/erb_no_unused_local_variable.rs
use std::collections::HashSet;

use crate::offense::UnboundOffense;
use crate::rule::{LintContext, ParserRule, Rule};
use crate::utils::prism_utils::{call_arguments, walk_prism};
use crate::utils::source_slice::location_from_offset;

use herb::prism::PrismNode;
use herb::ParseResult;
use herb_config::{Severity, SeverityConfig};

const IGNORED_PREFIX: &str = "_";

const REFERENCING_WRITES: &[&str] = &["LocalVariableOperatorWriteNode", "LocalVariableAndWriteNode", "LocalVariableOrWriteNode"];

pub struct ERBNoUnusedLocalVariableRule;

impl Rule for ERBNoUnusedLocalVariableRule {
  fn name(&self) -> &'static str {
    "erb-no-unused-local-variable"
  }

  fn introduced_in(&self) -> Option<&'static str> {
    Some("unreleased")
  }

  fn default_severity(&self) -> SeverityConfig {
    SeverityConfig::PerMode {
      cli: Severity::Error,
      editor: Severity::Info,
    }
  }

  fn parser_options(&self) -> herb::ParserOptions {
    herb::ParserOptions {
      prism_program: true,
      ..crate::rule::default_linter_parser_options()
    }
  }
}

#[derive(Default)]
struct LocalVariableCollector {
  referenced: HashSet<String>,
  argument_writes: HashSet<usize>,
}

impl LocalVariableCollector {
  fn collect(&mut self, program: &PrismNode) {
    walk_prism(program, &mut |node| {
      if node.is("CallNode") {
        for argument in call_arguments(node) {
          if argument.is("LocalVariableWriteNode") {
            self.argument_writes.insert(argument.start_offset);
          }
        }
      }

      if node.is("LocalVariableReadNode") || REFERENCING_WRITES.contains(&node.node_type.as_str()) {
        if let Some(name) = node.name.clone() {
          self.referenced.insert(name);
        }
      }

      true
    });
  }
}

fn advice_for(collector: &LocalVariableCollector, write: &PrismNode, name: &str) -> String {
  if collector.argument_writes.contains(&write.start_offset) {
    return format!(
      "This assignment sits in an argument list, where it still passes the value positionally. Write `{name}:` if you meant a keyword argument, or drop the `{name} =`."
    );
  }

  format!("Remove the assignment, or prefix it with an underscore as `_{name}` to show it is intentionally unused.")
}

impl ParserRule for ERBNoUnusedLocalVariableRule {
  fn check(&self, result: &ParseResult, context: &LintContext) -> Vec<UnboundOffense> {
    let source = if context.source.is_empty() { &result.source } else { &context.source };

    let program = match result.value.prism() {
      Some(program) if !source.is_empty() => program,
      _ => return Vec::new(),
    };

    let mut collector = LocalVariableCollector::default();

    collector.collect(program);

    let mut offenses = Vec::new();

    walk_prism(program, &mut |node| {
      if !node.is("LocalVariableWriteNode") {
        return true;
      }

      let name = match node.name.as_deref() {
        Some(name) => name,
        None => return true,
      };

      if name.starts_with(IGNORED_PREFIX) || collector.referenced.contains(name) {
        return true;
      }

      let (start, end) = match node.field_location("name_loc") {
        Some(span) => span,
        None => return true,
      };

      offenses.push(UnboundOffense::with_tags(
        self.name(),
        format!("Local variable `{name}` is assigned but never used. {}", advice_for(&collector, node, name)),
        location_from_offset(source, start, end),
        vec!["unnecessary".to_string()],
      ));

      true
    });

    offenses
  }
}
